error-prone-contrib/src/main/java/tech/picnic/errorprone/bugpatterns/CanonicalConstantFieldName.java

```diff
@@ -0,0 +1,153 @@
+import com.sun.source.tree.Tree;
+import com.sun.source.tree.VariableTree;
+import com.sun.source.util.TreeScanner;
+import java.util.Locale;
+import java.util.regex.Pattern;
+import javax.inject.Inject;
+import javax.lang.model.element.Modifier;
+import org.jspecify.annotations.Nullable;
+import tech.picnic.errorprone.utils.Flags;
+
+/**
+ * A {@link BugChecker} that flags constant variables that do not follow the upper snake case naming
+ * convention.
+ *
+ * <p>This check will rewrite the following variables with all its references:
+ *
+ * <pre>{@code
+ * private static final int simpleNumber = 1;
+ * }</pre>
+ *
+ * <p>To the following:
+ *
+ * <pre>{@code
+ * private static final int SIMPLE_NUMBER = 1;
+ * }</pre>
+ *
+ * @apiNote This check has an optional flag `ExcludedConstantFieldNames` that represents a list of
+ *     field names to exclude from this check.
+ */
+@AutoService(BugChecker.class)
+@BugPattern(
+    summary = "Constant variables should adhere to the `UPPER_SNAKE_CASE` naming convention",
+    link = BUG_PATTERNS_BASE_URL + "CanonicalConstantFieldName",
+    linkType = CUSTOM,
+    severity = WARNING,
+    tags = LIKELY_ERROR)
+public final class CanonicalConstantFieldName extends BugChecker implements VariableTreeMatcher {
+  private static final long serialVersionUID = 1L;
+  private static final Matcher<Tree> IS_CONSTANT =
+      allOf(
+          hasModifier(Modifier.STATIC), hasModifier(Modifier.PRIVATE), hasModifier(Modifier.FINAL));
+  private static final Pattern TO_SNAKE_CASE = Pattern.compile("([a-z])([A-Z])");
+  private static final ImmutableSet<String> DEFAULT_EXCLUDED_CONSTANT_FIELD_NAMES =
+      ImmutableSet.of("serialVersionUID");
+  private static final String EXCLUDED_CONSTANT_FIELD_NAMES =
+      "CanonicalConstantFieldName:ExcludedConstantFieldNames";
+
+  private final ImmutableList<String> optionalExcludedConstantFieldNames;
+
+  /** Instantiates a default {@link CanonicalConstantFieldName} instance. */
+  public CanonicalConstantFieldName() {
+    this(ErrorProneFlags.empty());
+  }
+
+  /**
+   * Instantiates a customized {@link CanonicalConstantFieldName}.
+   *
+   * @param flags Any provided command line flags.
+   */
+  @Inject
+  CanonicalConstantFieldName(ErrorProneFlags flags) {
+    optionalExcludedConstantFieldNames = getAllowedFieldNames(flags);
+  }
+
+  @Override
+  public Description matchVariable(VariableTree tree, VisitorState state) {
+    String variableName = tree.getName().toString();
+    if (IS_CONSTANT.matches(tree, state)
+        && !isUpperSnakeCase(variableName)
+        && !isVariableNameExcluded(variableName)) {
+      SuggestedFix.Builder fixBuilder = SuggestedFix.builder();
+
+      ImmutableList<VariableTree> variablesInCompilationUnit =
+          getVariablesInCompilationUnit(state.getPath().getCompilationUnit());
+      String replacement = toUpperSnakeCase(variableName);
+      if (variablesInCompilationUnit.stream()
```

Comment:
Just a suggestion but what if we extract this to a method, I think it reads nicely. Additionally if we use `anyMatch` instead of `noneMatch` it reads as "isVariableNameInUse`, instead of the negated variant, could be a bit easier to understand.

Reply:
More readable, I agree 💯